### PR TITLE
Add Python client for Project deployer

### PR DESCRIPTION
**Pull request by AgatheG**
[ch47130](https://app.clubhouse.io/dataiku/story/47130/public-api-for-project-deployer)

See PR for the Public API on dip: https://github.com/dataiku/dip/pull/11101

### CODE REVIEW

File: dataikuapi/dss/projectdeployer.py [hunk 1]
```diff
@@ -0,0 +1,450 @@
+from .future import DSSFuture
+
+class DSSProjectDeployer(object):
```
Comment (mhoarau) [inline dataikuapi/dss/projectdeployer.py:3]: Just a few code style issues: here we miss a blank line, later we have spaces around a `=` for default values, ...

File: dataikuapi/dss/projectdeployer.py [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+        if origin_project_key is None and design_node_url is None and design_node_id is None:
+            params = None
+        else:
+            params = {
+                "projectKey": origin_project_key,
+                "nodeId": design_node_id,
+                "url": design_node_url
+            }
+        return self.client._perform_empty("POST",
+                "/project-deployer/projects/%s/packages" % (self.project_key), params=params, files={"file":fp})
```
Comment (mhoarau) [inline dataikuapi/dss/projectdeployer.py:380]: The correct endpoint is `bundles`, not `packages`.

File: dataikuapi/dss/projectdeployer.py [hunk 3]
```diff
@@ -0,0 +1,450 @@
+from .future import DSSFuture
```
Comment (mhoarau) [inline dataikuapi/dss/projectdeployer.py:1]: Is it on purpose that we don't expose:
* `com.dataiku.dip.server.api.PublicAPIProjectDeployerController#deleteProject`
* `com.dataiku.dip.server.api.PublicAPIProjectDeployerController#listStages` 

?

**Review by mhoarau — CHANGES_REQUESTED** [review]

File: dataikuapi/dss/projectdeployer.py (hunk 3 above)
Comment (AgatheG) [reply dataikuapi/dss/projectdeployer.py:1]: For now I merely followed the methods we had in `apideployer.py`, "translating" them to Project deployer methods. For `deleteProject` especially, there was no endpoint in the Public API to delete services before dataiku/dip#11101. We can consider adding those to both deployers? 🙂

File: dataikuapi/dss/projectdeployer.py (hunk 1 above)
Comment (AgatheG) [reply dataikuapi/dss/projectdeployer.py:4]: Should be OK after ab1dfc1, if I missed some please say so 😁

**Review by instanceofme — APPROVED** [review]

File: dataikuapi/dss/projectdeployer.py [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+        if origin_project_key is None and design_node_url is None and design_node_id is None:
+            params = None
+        else:
+            params = {
+                "projectKey": origin_project_key,
+                "nodeId": design_node_id,
+                "url": design_node_url
+            }
+        return self.client._perform_empty("POST",
+                "/project-deployer/projects/%s/bundles" % (self.project_key), params=params, files={"file":fp})
```
Comment (marchelleboid) [inline dataikuapi/dss/projectdeployer.py:381]: If you try to send any of the optional params, they won't be received on the DSS side: `designNodeInfo` will be `null` as the controller is looking for a request param with that name. One way to fix this is by doing `params={"designNodeInfo": json.dumps(params)}` (and `import json` at the top), which leads to a pretty ugly request. Maybe it's better to change the controller method to accept each of those three optional params individually? In that case, this code could remain unchanged, you would just need to build the `DesignNodeInfo` object manually in the controller method in the public API PR.

**Review by marchelleboid — CHANGES_REQUESTED** [review]

File: dataikuapi/dss/apideployer.py [hunk 5]
```diff
@@ -10,7 +11,7 @@ class DSSAPIDeployer(object):
     def __init__(self, client):
         self.client = client
 
-    def list_deployments(self, as_objects = True):
+    def list_deployments(self, as_objects=True):
```
[the remark this answers is not in the document]
Comment (AgatheG) [reply dataikuapi/dss/apideployer.py:14]: Nice catch!

File: dataikuapi/dss/projectdeployer.py (hunk 4 above)
Comment (AgatheG) [reply dataikuapi/dss/projectdeployer.py:379]: Done in c1ed433, corresponding changes on the Public API endpoint in a86ba14

File: dataikuapi/dss/projectdeployer.py (hunk 3 above)
Comment (AgatheG) [reply dataikuapi/dss/projectdeployer.py:1]: Done in 6a220bd

**Review by marchelleboid — APPROVED** [review]
LGTM (only comment is the naming of the params on the other PR)

**Review by mhoarau — APPROVED** [review]
LGTM (I did not extensively played with all the APIs though)

**Pull request merged**